### PR TITLE
perf: Avoid creating a SmallVec if nothing changes during a fold

### PR DESCRIPTION
**Pull request by Marwes**
Not sure if this helps but in theory it should be less work than what
the current micro optimization does for `ty::Predicate` lists.

(It would explain the overhead I am seeing from `perf`.)

### CODE REVIEW

File: src/librustc/ty/structural_impls.rs [hunk 1]
```diff
@@ -1073,3 +1060,29 @@ impl<'tcx> TypeFoldable<'tcx> for InferConst<'tcx> {
         false
     }
 }
+
+fn fold_list<'tcx, F, T>(
```
Comment (jonas-schievink) [inline src/librustc/ty/structural_impls.rs]: Might be worth adding a comment here that explain that this is an optimization over `.iter().map(...).collect()`

**Comment by Marwes** [comment]
This could use a perf run now.

File: src/librustc/ty/structural_impls.rs (hunk 1 above)
Comment (Marwes) [reply src/librustc/ty/structural_impls.rs]: Done in https://github.com/rust-lang/rust/pull/68031/commits/a1586f1d2b17d687444d3b94aedd7ce24ae074ed

**Comment by Marwes** [comment]
Perf results look good. Ready to merge?

**Pull request merged**